### ISSUE TITLE
Pure functions return values

### ISSUE THREAD

**Issue by dbalabka** [vimeo/psalm#3427]
Is it possible to memorize the returned type of pure functions and use it in the following analysis: https://psalm.dev/r/30ee2f996a ?

Let's have the following pure function with nullable return type:
```php
<?php
 /**
 * @psalm-pure
 */
function getNullableInt(bool $return): ?int;
```
also we have following function that accepts only not nullable `int` type:
```php
<?php
function printInt(int $int): void;
```

I assume that the following code should not throw an error:
```php
<?php
if (getNullableInt(true)) {
    printInt(getNullableInt(true));
}
```
```
ERROR: PossiblyNullArgument - 30:14 - Argument 1 of printInt cannot be null, possibly null value provided
```

I understand that most of such cases should be handled with additional variable, but not all our code is ideal and sometimes it contains the naive implementation. IMO Psalm should be more flexible and not report such cases, by forcing the developer to change the code.
```php
// right way
$int = getNullableInt(true);
if ($int) {
    printInt($int);
}
```

**Comment by dbalabka** [comment]
Here is a more complicate example with mutation-free class methods:
https://psalm.dev/r/0a100c3253

**Comment by muglug** [comment]
the `memoizeMethodCall` flag was conflicting with the property proxying that @m0003r added

**Comment by muglug** [comment]
Sorry, to be clear that just fixes the method memoisation. I'm not going to fix the function memoisation, though – I think it'll add more complexity than necessary.

**Comment by dbalabka** [comment]
@muglug agree with you. Thanks for the quick reply!
